i18n: Fix bug where package-level variables are not translated.

brianpursley opened the pull request:
#### What type of PR is this?
/kind bug

#### What this PR does / why we need it:

The call to `i18n.LoadTranslations()` needs to occur on init so that package-level variables that call `i18n.T()` to initialize their values will be able to receive the translated string.

Added new integration tests to test help output translation.

#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes https://github.com/kubernetes/kubectl/issues/1327

#### Special notes for your reviewer:

I *think* it should be OK to move `LoadTranslations` to `init`, but let me know if you can think of any problems this would cause.  The alternative to this would be we say you cannot use `i18n.T()` to initialize package-level variables.

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

brianpursley commented:
/hold for review

KnVerey commented on staging/src/k8s.io/kubectl/pkg/util/i18n/i18n.go:
```diff
@@ -52,6 +52,11 @@ var knownTranslations = map[string][]string{
 
+func init() {
+	// LoadTranslations must occur in init so that package-level variables initialized with strings can be translated
+	LoadTranslations("kubectl", nil)
```
A possible issue that occurs to me with this approach is library usage. In theory, a library consumer constructing their own commands might want to supply their own `getLanguageFn` (`func LoadTranslations(root string, getLanguageFn func() string) error `), or not load translations at all.

Perhaps an alternative could be to have `i18n.T` itself lazy-load the translations? If there isn't an easy way to detect whether they're already loaded, `LoadTranslations` itself, or a wrapper charged with managing this, could set one?

brianpursley replied:
I added lazy loading.

I exposed it through a `SetLoadTranslationsFunc()` function that can be used by libraries if they want to substitute their own lazy loading function.  They could still call `LoadTranslations` too.

Its a little more complicated than I was going for initially, but I think it should work.

brianpursley commented on staging/src/k8s.io/kubectl/pkg/util/i18n/i18n.go:
```diff
@@ -52,6 +54,22 @@ var knownTranslations = map[string][]string{
+// SetLoadTranslationsFunc sets the function called to lazy load translations.
+// It should be called in an init() func.
+func SetLoadTranslationsFunc(f func() error) error {
+	if translationsLoaded {
```
I was thinking they would just call `LoadTranslations` but they don't have to, so yeah, might as well set it for them here too.  I made the change.

KnVerey replied:
👍 can we add a test for this?

brianpursley replied:
Added a test and updated existing tests to check translationsLoaded actual vs expected

brianpursley commented:
In case anyone is wondering how overriding the translations would work for a third party, you will have to be VERY EARLY in the init process to make sure it occurs before *any* call to `i18n.T()`.

The trick is you have to create a separate package with just an init func, like this:
```
package myi18n

import "k8s.io/kubectl/pkg/util/i18n"

func init() {
	if err := i18n.SetLoadTranslationsFunc(loadCustomTranslations); err != nil {
		panic(err)
	}
}

func loadCustomTranslations() error {
	// Do your custom load here...
}
```

Then (and here is the critical part), you need to import it *before* any other packages that might call `i18n.T()`, which is basically any kubectl command.  This is one of those cases where the order of the imports matters.

And since it has no functions in this package, you need to import it for side effects, like this:

```
import (
...
        _ "example.com/myapp/myi18n"
...
)
```

There might be another way, but this was the only way I could force the init to be early enough so that it could occur before any translations are attempted.

brianpursley commented:
/retest

brianpursley committed:
i18n: Fix bug where package-level variables are not translated.

Change i18n.T() to load translations if they have not yet been loaded.

Added new integration tests to test help output translation.

brianpursley commented on staging/src/k8s.io/kubectl/pkg/util/i18n/i18n.go:
```diff
@@ -52,6 +54,30 @@ var knownTranslations = map[string][]string{
+)
+
+// SetLoadTranslationsFunc sets the function called to lazy load translations.
+// It should be called in an init() func.
```
Updated the comment to include the details I put in the PR comment

KnVerey commented:
/lgtm

brianpursley commented:
/unhold

dims commented:
/retest
